### DISCUSSION TITLE
Path or Expectation of FJS

### DISCUSSION THREAD

**Discussion by climba03003** [fastify/fast-json-stringify#532] [Polls]
Within the discussion inside https://github.com/fastify/fast-json-stringify/pull/530.
I believe we are lost on the path or expectation of this module.

It is clear that we divide into two separate opinion.

1. FJS should be a plain serializer, we shouldn't care on the validation.

2. FJS should be a serializer that return the `value` can be validate against the `JSON-Schema`.

I open a poll here to helps us make decision on which path should we go.

**Comment by ivan-tymoshenko** [comment]
I want to discuss with you how FJS should work. Especially in terms of validation and output guarantees. Now FJS is a serializer with a mixin of validation checks. Some of these checks are required to define what to do with data (oneOf, anyOf, if/else). Some of these checks are not needed for serialization and check basic validation rules.

There are two different points of how FJS should work.
1. FJS doesn't do any validation except in those cases when it's inevitable. That means if you want to get the correct output, you should pass the correct/validated input.
```js
const schema = {/*json schema*/}
const data = {/*input data*/}

// If you pass correct data it will work correctly
if (validate(data, schema)) {
  const serializedData = FJS(data, schema)
  const parsedData = JSON.parse(serializedData)
  assert(validate(parsedData, schema)) // should always be true
} 

// If you don't know that you input is valid, there is no guarantee that FJS produce valid output
const serializedData = FJS(data, schema) // might throw if can't produce the result, but doesn't have explicit validation checks
const parsedData = JSON.parse(serializedData)
assert(validate(parsedData, schema)) // unkown
```

2. FJS should always produce output that matches the schema.
```js
const schema = {/*json schema*/}
const data = {/*input data*/}

const serializedData = FJS(data, schema) // if can't produce the correct (that will valid to the schema) output it will throw an error
const parsedData = JSON.parse(serializedData)
assert(validate(parsedData, schema)) // true
```

If you have any other ideas, you can share them.

**Reply by Eomm** [reply]
I think we could adapt JSON Schema to serialization too.

For example: if the required field is missing, we will fall back to the `default` property.. that could be `null`, that is still a valid JSON value

**Reply by ivan-tymoshenko** [reply]
I think it's a bad idea because you will not know if it's really a null value or undefined. It's ok to throw an error if we can't produce valid output. In case we want to guarantee that output is valid.

**Reply by Eomm** [reply]
Let's try to speak about the structure instead of the fields' value:

- the validator validates the structure and the data
- the serializer applies a structure and does not validate the data

do you agree with that?

**Reply by ivan-tymoshenko** [reply]
Yes

**Reply by Eomm** [reply]
So is a property a part of the structure or of the output value? 😄

I think it is the former

**Reply by ivan-tymoshenko** [reply]
Ok, what about data? We should serialize it somehow. What do you expect from the output data?

IMHO nullable is an interesting case to discuss it. Why do we coerce null to “”, 0 or false, if it’s a not nullable string, number or boolean. Why we can't keep it null, if serializer doesn't care about data.

**Reply by Eomm** [reply]
Usually, I expect the type is consistent with the one set on the schema e I have only the defined properties (fields filtering)

> Why we can't keep it null, if serializer doesn't care about data.

I think `nullable` is an argument of the hypothetical json schema field's `serialize()` function

**Reply by ivan-tymoshenko** [reply]
Can you describe what "type is consistent with the one set on the schema" means? That is the main question. Does that mean that the output value should always be valid to the schema? Or it means something else?

> I think nullable is an argument of the hypothetical json schema field's serialize() function

I'm sorry, I didn't get your point. I want to discuss nullable because it's a useful example where we can see what guarantees we get on the output.

I will give you an example:
1. You have a schema like that:
```js
{ type: 'string' }
```
What can you tell about the output of serialization? Well, if you pass a string, you will get a string, it's pretty obvious. If you pass a value of another type, the value would be coerced, and you will get a string. If you pass the null value, you will get an empty string. So it looks like you always get valid output.

2.
```js
{ type: 'string', const: 'foo' }
```
It doesn't matter what you pass, you will always get the 'foo' value. It also looks like you always get valid output.

3. Here is a more complicated one
```js
{ type: 'string', format: 'date' }
```
What should we return if you pass the null value in this case? The empty string is not valid to this schema. It must be a date. Technically to be consistent, we should return a string with some default date, but it's a bad idea. Should we throw an error, or what?

In other words, what is a set of possible output values for schema `{ type: 'string', format: 'date' }`? Is it only a date string? Or can it still be any string?

**Reply by Eomm** [reply]
> Or it means something else?

I mean that the `type` attribute has priority over all other options.

So, talking about your example:

> Is it only a date string? Or can it still be any string?

I think `{ type: 'string', format: 'date' }` should be an empty string.

Reading [the spec](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-7.1):


>  If the type of the instance to validate is not in
   this set, validation for this format attribute and instance SHOULD
   succeed.

If I translate this sentence to a serializer, I think that when the input value does not apply to the format, the format is ignored (aka succeed).

Contrary, I think we are talking about defining a new JSON-Schema for the serializer.

**Reply by ivan-tymoshenko** [reply]
@Eomm Well, if `format` is an ambiguous example (Ajv would validate an empty string false with the date format). There is another example:
```js
{ type: 'string', enum: ['foo', 'bar'] }
```
How should we handle `null` case? Or any other string except 'foo', 'bar'?

**Reply by ivan-tymoshenko** [reply]
@Eomm Can you please answer this question, when you have some time. I don't want to leave this discussion unresolved.

**Reply by Eomm** [reply]
Sorry for the delay, but I came across a proposal.

I would not focus on Serialization Process **VS** Validation Phase but on a few pillars:

1. as a serializer, I will never reject an input
2. during the serialization, if I don't understand the JSON schema instructions, I will fallback to clear and defined defaults
3. optionally, when I don't know how to serialize the input, I will throw an error

> How should we handle null case? Or any other string except 'foo', 'bar'?

This is unpredictable: we will fallback to the `type:string` default value. In case the `strict` mode is on, we would throw an error

**Reply by ivan-tymoshenko** [reply]
> as a serializer, I will never reject an input

I don't understand what does it mean. It's vague, why it should reject the object without required field then? Can you specify what does it mean.

> if I don't understand the JSON schema instructions, I will fallback to clear and defined defaults

What do you mean by "I don't understand the JSON schema instructions". If there is something wrong in the schema we should fail on compile time.

> optionally, when I don't know how to serialize the input, I will throw an error

Here I 100% agree. If we don't know how to coerce the data we should fail.

> This is unpredictable: we will fallback to the type:string default value. In case the strict mode is on, we would throw an error

Ok I want to continue with this example, because I think it can give an answer to the main question. If there is a schema with one enum item, what should we do if we receive null or another value?
```js
{ type: 'string', enum: ['foo'] }
```
Should we also fallback to the `type:string` value?

**Reply by Eomm** [reply]
> I don't understand what does it mean.

I mean that the default behaviour should be to serialize even if the input does not match at all with the schema

> What do you mean by "I don't understand the JSON schema instructions"

I mean those cases where the Schema tells me to do something that the input does not apply to.
The `{ type: 'string', enum: ['foo'] }` example will `input = null` is one of them.

> Should we also fallback to the type:string value?

In a `strictMode: false`: Yes, exactly
In a `strictMode: true`: No, we should throw

**Reply by ivan-tymoshenko** [reply]
> I mean that the default behaviour should be to serialize even if the input does not match at all with the schema

According to this sentence we shouldn't throw an error if there is no required field. Can you specify you thought more precisely.

> In a strictMode: false: Yes, exactly

```js
{ type: 'string', const: 'foo' }
```

Ok, if I have `const` string schema and receive `bar`. Should I serialize it as `bar`?

> In a strictMode: true: No, we should throw

We don't have strictMode support right now. Let's talk first about current/default implementation.

**Reply by Uzlopak** [reply]
JSON.stringify("foo") does not serialize to "bar"

**Reply by ivan-tymoshenko** [reply]
> JSON.stringify("foo") does not serialize to "bar"

I understand your point. If we have const string in schema we should not substitute the const value instead of any string. It make sence, but you have a wrong argument. FJS doens't work as JSON.stringify, so I can make a lot of examples like that.
```
JSON.stringify(null) doesn't serialize to 0
JSON.stringify(/.*/) doesn't serialize to ".*"
JSON.stringify(32.33) doesn't serialize to 32
JSON.stringify(555) doesn't serialize to true
```
But all these example are valid with different schemas.

**Reply by ivan-tymoshenko** [reply]
Okay, we can assume that FJS should not give you any guarantees on the output and it might use the schema as a "hint" to speed up serialization. Then I again have a question. Why should it throw an error if there is no required field?

DISCLAIMER: I understand that it's an important check for Fastify and a lot of people relyes on it. I don't want to just remove it. But it doesn't mean that it inconsistent check for FJS.

@Eomm I read one more time your point about structure and data. I think it can be your point, but it's not "proof". I will explaint. 

Validation has two arguments `schema` and `data`. And it checks if data corresponds to the `schema`. Usually it checks both types and structure. Here we all agree I guess.

Serializer in general has one argument `data` and converts it to the string. In general it has only one rule: serialized data should be able to be parsed by any parser that support this serialization format. Should JSON serializer preserve `data` argument structure? I think yes. Should serializer validate if `data` structure corresponds to another argument `schema`? I don't know, but it's definetly another question. And it's more that a serializer then, so we can't "proof" something using the definition if serializer.

**Reply by Eomm** [reply]
> According to this sentence we shouldn't throw an error if there is no required field

"It depends":
strict mode true: should throw as FJS "does not understand the JSON schema instructions"
strict mode false: should not throw

> Ok, if I have const string schema and receive bar. Should I serialize it as bar?

Yes, we should serialize as `bar` since FJS knows "understand the JSON schema instructions".

> We don't have strictMode support right now

Yes, there is not such feature; I though we was talking about what we would like to do in next major to be more consistent.

> that FJS should not give you any guarantees on the output and it might use the schema as a "hint" to speed up serialization

Yeah, I quite agree with that. The output precision should depends on the schema quality and FJS configuration

> Should serializer validate if data structure corresponds to another argument schema?

I wonder if we can open this decision to the developer

**Reply by ivan-tymoshenko** [reply]
> Yes, we should serialize as `bar` since FJS knows "understand the JSON schema instructions".

So that means that this [comment](https://github.com/fastify/fast-json-stringify/pull/505#issuecomment-1219727114) was wrong.
`I think that we should force the const value into the serialized object whatever the input object is.`
What I want to say is that it's easy to make a mistake in some edge cases if we don't have a strong vision/concept of how it should work.

> Yes, there is not such feature; I though we was talking about what we would like to do in next major to be more consistent.

Maybe I will change my mind, but I don't think it's possible. It's easier to write another package for this than maintain a strict and unstrict version in the same codebase. I understand your idea. Let's talk about the unstrict version now because it's a current working version. And I think we can't make a strict version as a default, even with a Fastify major.

> > According to this sentence we shouldn't throw an error if there is no required field

> "It depends":
strict mode true: should throw as FJS "does not understand the JSON schema instructions"
strict mode false: should not throw

I agree with this point of view. I'm not talking about the strict mode (see the previous comment). The problem is that Fastify uses FJS not just as a serializer but also as a validator for some rules like required fields. I understand the benefits, it's working pretty fast. But it creates a situation when we can't optimize FJS because we can't just remove those checks from Fastify (I 100% agree with that), and on the other hand, I, as a Fastify user, can't rely on this "validation". Is my serialized data valid to the schema after serialization? I don't know, there is no guarantee. I have to validate it to be sure.

**Reply by Eomm** [reply]
> So that means that this https://github.com/fastify/fast-json-stringify/pull/505#issuecomment-1219727114 was wrong.

I don't agree: I think the `const` keyword is a definition, so FJS applies the schema instruction.
If FJS throws an error when the input value is different from the `const` it would perform a validation process.

The main point I see here is that we can't apply the same rule for every keyword. Does it make sense?

Example:

- type: coerce the value
- const: coerce the value
- requires: (no strict mode) coerce the value using the type set (what if the type is an array with min 1??)

> Is my serialized data valid to the schema after serialization?

We are walking to the "we should throw if we can't serialize" side.
In this case, the impact would be to provide the Fastify users an "onSerializeError" to deal with because, let's think of a `post /create-todo`:

- validate input
- insert the todo
- process something
- serialize response
- error during serialization due a bug

=> what is the state of the endpoint now?

This is the main reason I'm PRO to the sentence:

> the serializer should never throw

because it is the last thing and all the business logic is done

**Reply by ivan-tymoshenko** [reply]
@Eomm Sorry for the _really_ long delay.

> I don't agree: I think the const keyword is a definition, so FJS applies the schema instruction.
If FJS throws an error when the input value is different from the const it would perform a validation process.

[Here](https://github.com/fastify/fast-json-stringify/discussions/532#discussioncomment-3836650) you said that if we have a value different from a constant we should serialize it and not substitute a const value. In this [PR](https://github.com/fastify/fast-json-stringify/pull/505#issuecomment-1219727114) we decided to substitute a const value whatever the input is. That is why I'm saying that it was a mistake. We don't have to through an error, though. We just need to implement another logic. Instead of:
```js
const constValue = 'bar'
const serializedValue = '"bar"' // precompiled serialized const value

function serializeConst (value) {
  return serializedValue // always returns const value
}
```
we should do
```js
const constValue = 'bar'
const serializedValue = '"bar"' // precompiled serialized const value

function serializeConst (value) {
  if (value === constValue) {
    return serializedValue // returns const only if input match the const. fast path
  }
  return serialize(value)
}
```

> The main point I see here is that we can't apply the same rule for every keyword. Does it make sense?

I agree that keywords in the json schema spec are not "homogeneous". And originally, they were defined for validation, so we have all these problems only because we are trying to define what all these keywords should mean in the serialization process. But we can't just say ok, all keywords are different let's pick randomly some of them and apply them somehow (I'm not joking, some part of the code has no explanation). There should be logic in these definitions/serialization rules. There could be exceptions, but they also should have a reason and explanation. 

> This is the main reason I'm PRO to the sentence:

>>the serializer should never throw

Sometimes we have a situation when we don't know what to do with this data. As an option, we can always fall to the default JSON.stringifty instead of throwing an error. But it also should be defined as a part of the concept. There is a statement that I prefer:

**fast-json-stringify guarantees that you will get a valid output only if your input matches the schema or can be coerced to the schema. If your input doesn't match the schema, you will get undefined behavior.**

This means that during the compilation, we are trying to build the serialization function that expects that you pass already valid data (or data that can be coerced). If an input matches the schema, we guarantee that you will get the correct output. If the input doesn't match the schema, you will get a UB. But, unfortunately, there is only me who likes it.

**Reply by Eomm** [reply]
> Here you said

I think there is a misunderstanding because I thought that `bar` was the `const` value. When I wrote `understand the JSON schema instructions` I meant "FJS can process the const keyword by setting the const value as output".

> we should do

The logic you describe is clear, but I don't agree: why the user has set `const`, if the wanted behaviour comes from `type: string`? I see some logic overlapping by implementing the proposed solution. Is this a matter of JSON schema reusability?

> fast-json-stringify guarantees that you will get a valid output only if your input matches the schema or can be coerced to the schema. If your input doesn't match the schema, you will get undefined behavior.

I like this definition. I don't think we should say an "undefined behavior", but a "possibly invalid json-string if validated with the schema". I keep saying that I'm pro to the no-throw pattern

**Comment by climba03003** [comment]
@Uzlopak Don't mind if I reply about the #520 case here.

`nullable` actually contradict with `const`. Inside the specification, I do not found anything about how to deal with the contradicting case, and which `keyword` should take higher priority than the others.
I think accepting #520 and reverting it because of user expectation is both valid consideration.

**Reply by Uzlopak** [reply]
I think #520 actually shows the dilemma were we are walking into. 

See https://github.com/fastify/fast-json-stringify/blob/1b9d5a1d52e4e4b96a64d431d1e8921951ee0ccf/test/const.test.js#L237 

So we would have stringifiying `{ "foo": null }` to `{ "foo": "baz" }` which would be imho plain wrong. But our argument is, that FSJ is a serializer and we later want maybe the validation. If I serialize `{ "foo": null }`  I want to get `{ "foo": null }` and not `{ "foo": "baz" }`, which the validator can later not determine was actually a invalid input. 

If I see it correctly #520 or a similar code change did not land in next, but stilll it makes we wonder.

Also regarding type and enum: If enum is set, it is basically intersectioning with the type. 

`{ "type": "string", enum: ["a"] }` => only "a" allowed
`{ "type": "string", enum: ["a", 2] }` => still only "a" allowed because 2 is not of type string

**Comment by Eomm** [comment]
I'm not sure which option I should vote.

Based on the json schema properties list there are edge cases, and filling out this matrix will explain how FJS should/will work

|keyword|is the keyword supported right now?|should the keyword be supported?|what to do when...|
|---|---|---|---|
`$id`|||
`$schema`|||
`$ref`|||
`$comment`|||
`title`|||
`description`|||
`default`|||
`readOnly`|||
`writeOnly`|||
`examples`|||
`multipleOf`|||
`maximum`|||
`exclusiveMaximum`|||
`minimum`|||
`exclusiveMinimum`|||
`maxLength`|||
`minLength`|||
`pattern`|||
`additionalItems`|||
`items`|||
`maxItems`|||
`minItems`|||
`uniqueItems`|||
`contains`|||
`maxProperties`|||
`minProperties`|||
`required`|||
`additionalProperties`|||
`definitions`|||
`properties`|||
`patternProperties`|||
`dependencies`|||
`propertyNames`|||
`const`|||
`enum`|||
`type`|||
`format`|||
`contentMediaType`|||
`contentEncoding`|||
`if`|||
`then`|||
`else`|||
`allOf`|||
`anyOf`|||
`oneOf`|||
`not`|||

**Reply by climba03003** [reply]
`$id`, `$schema`, `$comment`, `title`, `description`, `examples`, `contentMediaType`, `contentEncoding`.
These 8 `keyword` are metadata that do nothing on serialization or validation.
I think it can be removed first.

**Reply by ivan-tymoshenko** [reply]
This table doesn't work for me. I can fill this table in any way, depending on the way FJS should work. If it helps someone, you can fill it.

**Reply by Eomm** [reply]
It helps me to understand where FJS is and how to move forward, I will try to fill it in the next days

**Comment by jsumners** [comment]
Yes, this module is a standalone module usable by anyone to serialize data into a JSON representation. However, it is primarily developed to support Fastify. Fastify uses it to serialize response payloads by default and it is advertised to do so. The module should continue to provide the expectations that Fastify has around response serialization. That includes throwing an error when the data is missing required properties for the response to be valid.

**Reply by ivan-tymoshenko** [reply]
1. Why we can't do it in another module that Fastify will use? If we want to throw an error if there is no required property, why we can't validate the data before serialization?
2. It's easy to say let's build our module according to users' expectations. But it doesn't resolve the problem.

**Reply by ivan-tymoshenko** [reply]
I don't want to separate FJS from the Fastify use case. Quite the opposite, I want to optimize it as much as possible for this. But it doesn't mean that we shouldn't think about the library architecture.

**Reply by jsumners** [reply]
Your suggestion is:

```js
function routeHandler (req, reply) {
  const responseData = getDataFromSomewhere()
  if (this.validate(responseData) === false) { // does data validation
    return reply.code(500).send("we be broke")
  }

  // does data serialization which is effectively the same operation
  // as "data validation", thus doing the same operation twice in order
  // to send the response
  reply.send(responseData) 
}
```

**Reply by ivan-tymoshenko** [reply]
I don't agree with the comment, but yes this is one of the options.

**Reply by ivan-tymoshenko** [reply]
If you think it’s not the best way to do it, you can describe how you see it. Should we validate the incoming date in the serializer if it the same operation? Should we give the guarantee on the output? Describe how you see it should work.

**Reply by jsumners** [reply]
I'm rather done repeating myself. I stated my opinion in https://github.com/fastify/fast-json-stringify/pull/530 when @mcollina asked for my review and repeated it here in this sub-thread. I did not block the PR.

**Reply by ivan-tymoshenko** [reply]
Please forget about this PR. He is not the reason for this discussion. That PR is only one of the questions that we can't resolve without understanding where we are going. I got you. You want to be sure that if there is defined a required field in the schema, it should be in the output. It's ok. What other guarantees except required fields do you want to get on the output based on the scheme?

**Reply by Uzlopak** [reply]
The roundtrip is questionable but it would make sense to separate the concerns, as complex cases could get out of hand. 

On the other side we provide the schema to FSJ. And actually the schema tells how the shape should be. And we do parse anyway the json schema. So we can handle also the validation cheaper than using ajv afterwards. Fail early. Fail fast.

**Reply by ivan-tymoshenko** [reply]
@Uzlopak So you think that FJS should validate the input data?

**Reply by Uzlopak** [reply]
Kind of, yes

**Reply by ivan-tymoshenko** [reply]
@Uzlopak, as I understand, you would like to have the second option from here: https://github.com/fastify/fast-json-stringify/discussions/532#discussioncomment-3609675. If serialization didn't fail, then the output is valid/match to the schema.

We can do this. So that means that #520 is correct in that case. And cases like that https://github.com/fastify/fast-json-stringify/blob/master/test/date.test.js#L268 should throw an error because they are invalid. Did I understand you correctly?

**Reply by Uzlopak** [reply]
Regarding #520 : It is plain wrong to coerce null to a string.

**Reply by ivan-tymoshenko** [reply]
If you pass null to the FJS as not nullable string value, it will coerce it to empty string.

Otherwise it would be invalid output.

https://github.com/fastify/fast-json-stringify/blob/3539f2d3fc5795e0314bb79ef9b6ba1ff190fb8d/lib/serializer.js#L90

**Reply by ivan-tymoshenko** [reply]
In the same way we already coerce thanks to 0 and false for number and boolean type.

https://github.com/fastify/fast-json-stringify#nullable

**Reply by Uzlopak** [reply]
Well that makes maybe some sense. But what #520 does is transforming null to the `const` value, which is basically making `const` behave like what i would expect from the jsonschema `default` field.

**Reply by ivan-tymoshenko** [reply]
It makes the output of FJS valid to schema. If you try to valid thanks value with schema that consist const you will get false.

**Reply by ivan-tymoshenko** [reply]
No it’s not working as a default value. Default value will be substituted if there is no value. Null is not the absence of value.

Const will not be substituted if you decide to not pass the value.

**Reply by Uzlopak** [reply]
https://json-schema.org/understanding-json-schema/reference/generic.html#constant-values

`The const keyword is used to restrict a value to a single value.`

So it is not part of the serialization but of the validation. So coercing { foo: null } to { "foo": "bar" } is wrong.

Another example:
https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values

`The enum keyword is used to restrict a value to a fixed set of values. It must be an array with at least one element, where each element is unique.`

So you could say that const is a special case of enum. Enum with only one element can be written as const. If you use enum to which value would you coerce? And does this mean that i can avoid the coercion only by using enum with one element instead of const?

**Reply by ivan-tymoshenko** [reply]
Yes, you are right, and there are a lot of examples like that. That's exactly why we have this discussion. How would you like to handle the null value in the string, const, and enum cases?

**Reply by Uzlopak** [reply]
First of all: If it is type: ["string", "number"] and const is set to null then it is plain impossible to serialize the object to valid json. It is the strictness which is implied by the schema. Maybe we do some magical coercion, but then i would disable the coercion in my projects because i personally want no coercion of primitive types. 

And second: 
As i mentioned already above: The configuration is making it narrower, so if i use type: ["string", "null"] and const: "bla" than it means only the string "bla" is valid and the type "null" is basically ignored.

If i do `JSON.stringify({ foo: null })` and get `{"foo": null }` then i expect FSJ serializes it also to `{ "foo": null }` or fails to serialize because the schema is not matching. 

Gold standard here is JSON.stringify. If it deviates from the output of JSON.stringify then if is not fast json stringify but something else.

**Reply by ivan-tymoshenko** [reply]
>  i would disable the coercion in my projects

You can't disable coercion in FJS.

**Comment by mcollina** [comment]
fast-json-stringify was designed to be a _serializer_. However, I added some basic validation, like `required`, because it provides a good user experience.

Performing _full_ validation would slow it to a halt, make it useless. Some validation is also needed to support all the JSON Schema features.

**Comment by Uzlopak** [comment]
I made the PRs #611 and #579 to get some ground to talk. But to be frank, i am quite dissapointed as there is no clear goal to reach.

**Reply by Eomm** [reply]
[Tables](https://github.com/fastify/fast-json-stringify/discussions/532#discussioncomment-3609707) help in this case - I didn't fulfil yet as I don't have a fresh mind with the actual source code

**Comment by ivan-tymoshenko** [comment]
It might be good to design a pluggable compiler. So anyone could write his own plugin and inject some code into the precompiled serialization function. After that, we could insert into the code any validation or type coercion without modifying the compiler core module.

```js
const compiler = new Compiler({
  preArraySerializationHook: (ctx, array, arraySchema) => {
    if (array.length !== arraySchema.lenght) {
      throw new Error ('Array has wrong length')
    }
  },
  preIntegerSerializationHook: () => {},
  preObjectSerializationHook: () => {},
  // ...
  preSchemaSerializationHooks: {
    'schema-1': () => {},
    'schema-2': () => {},
  }
})

const serialize = compiler.build(schema, {
  // pre serialization hooks only for this schema
})

serialize(object)
```

P.S. I'm not sure that it's possible to write it, but I would try.

**Comment by autopulated** [comment]
Currently:

If we have a schema: `{type: 'object', properties: { a: { type: 'string' } } }`, and value `{a:123}`
then currently FJS will coerce the number to a string in the serialisation: `{a:"123"}`

If  we have a schema: `{type: 'object', properties: { a: { type: 'number' } } }`, and value `{a:"123"}`
then currently FJS will coerce the string to a number in the serialisation: `{a:123}`

If  we have a schema: `{type: 'object', properties: { a: { type: 'number' } } }`, and value `{a:new Date(123)}`
then currently FJS will coerce the date to a string in the serialisation: `{a:"1970-01-01T00:00:00.123Z"}` (is this just a bug?)

If we have a schema: `{type: 'object', properties: { a: { type: 'number' } } }`, and value `{a:{}}`
then currently FJS will throw that the object value cannot be converted to a number.

(`additionalProperties: false` in the schema does not change these results)

This existing behaviour seems inconsistent and surprising. I don't think fast-json-stringify should ever do a type coercion that `JSON.stringify` would not do (number->string or string->number). After all, this is `fast-json-stringify`, it is not `fast-json-schema-coercion`, nor is it `fast-json-schema-validation-then-stringification`.

That is, I think that the design goal of FJS should be a new option 3:

 * FJS should be a serializer where the result is the equivalent of `JSON.stringify(value)`,  as long as value matches the schema, or **a sub-set of what would be returned by `JSON.stringify(value)` otherwise**.
 
Properties missing from the value that are in the schema should always be omitted (even if the schema defines a default or const), and the result will always match the parts of the schema for which data is returned.

I don't think FJS should ever throw on a value that does not match the schema (including in the cases where it currently does throw, like providing an object for a number). If you want full validation, just validate your value with AJV before stringifying it.

But this raises a problem with anyOf/allOf/if-else: in these cases. For example, if we have: schema:
```
const serialiser = build({
    anyOf: [
        {type:'object', properties:{a:{type:'string'}, b:{type:'number'} }, additionalProperties:false},
        {type:'object', properties:{a:{type:'string'}, c:{type:'boolean'} }, additionalProperties:false}
    ]
})
```
What should `serialiser({a:'s', b:true, d:'extra'})` be?  (currently this will throw at serialisation-time).

Serialising these constructs is always going to be slower, because we need to check which option to proceed with before doing the serialisation. So the easy option would just be to give up and call `JSON.stringify`, but in that case the guarantee about not returning data that does not match part of the schema is gone. The other option is to validate and throw on incompatible values, but adding checks in all the simple cases to throw in a consistent manner when incompatible data is passed would dramatically slow things down.

I think it would be reasonable to instead restrict the ways that anyOf, oneOf, allOf, if-else can be used in schemas that are to be used with fast-json-stringify:

 * `allOf`: should just not be supported. This provides no useful information to enable fast serialisation. If you wanted to validate your data, and not serialise it, just use AJV.
 * `anyOf` and `oneOf`: must use a discriminator keyword extension, which is useful, fast-ish, and common in APIs. In other word, the serialiser can check which sub-schema to proceed with based on checking the value of a single property against a constant.
 * `if-else`: should probably also not be supported, or could support a discriminator in a similar manner to oneOf

These properties of the schema should be checked at compile time, and ignored or throw, depending on a strict option. If they are ignored, the corresponding parts of a schema would be omitted from serialisation (not just delegated to JSON.stringify). So for example a property with a oneOf subschema without a discriminator keyword would always be omitted from serialisation, regardless of the value passed, or would throw at compile time if compiled with a strict option.

I think this approach would provide an easy to understand, consistent, fast, and safe* way of serialising data, and clearly defines what fast-json-stringify does (serialisation), and does not do (validation of the data to be serialised).

*The safety that I think matters most in the context of serialising data for an API is making sure that only data that the programmer intends to be serialised is serialised (preventing unexpected data leaking in an API).


I appreciate that this approach would be a big change from the current behaviour, but having poked around fast-json-schema a bit recently, I thought this view of things might be useful, and this thread seemed like the place for this discussion.

**Reply by ivan-tymoshenko** [reply]
@autopulated Thanks for your feedback. Before discussing implementation details we should understand can we do this and if yes, then how we can do this. I aslo would like to have a consistent and simple idea of how this library should work. Your idea of making it work exactly as JSON.stringify is consistent and simple, although I don't think it will work for a coumple of reasons.

1. It would be really hard to make it work as an optional behavior as it impacts almost all core serialization rules we have. IMHO it would be easier to fork/clone the project and rewrite it than keep everything in a same code base.
2. We can't add a big feature that is not used in fastify by default, because we don't have anough people/time to mainain it. Don't understand me wrong, pr is always welcome, but it's highly possible that it would require a lot of work and wouldn't bring a lot of benefits.
3. We can't make your concept a default behavior because it would be a significant breaking change and because a lot of core mainainers don't want this library to work in this way.

**Reply by autopulated** [reply]
No worries, thanks for explaining, totally understand all of your points.

**Comment by mcollina** [comment]
After a few years and growing a bit more gray hair, I think you were right and this module should not care about validation. This should enable a few improved perf things.
